src/coreclr/src/jit/morph.cpp

```diff
@@ -13034,6 +13034,30 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
                 op2 = tree->AsOp()->gtOp2;
             }
 
+            // See if we can fold floating point operations
+            if (varTypeIsFloating(tree->TypeGet()) && !optValnumCSE_phase)
+            {
+                if ((oper == GT_MUL) && !op1->IsCnsFltOrDbl() && op2->IsCnsFltOrDbl())
+                {
+                    if (op2->AsDblCon()->gtDconVal == 2.0)
```

Comment:
In https://github.com/dotnet/coreclr/pull/24584 we introduced an optimization which replaces `x / pow2` with `x * (1 / pow2)`.

In the case `pow2` is `0.5`, this would transform `x / 0.5` into `x * 2.0` and so the question is: Will this still be correctly optimized in this scenario? That is, will it become `x + x`?

Reply:
@tannergooding heh, yeah, I checked that 🙂 

`x / 0.5`
is optimized to `vaddss`

Reply:
Do we have a general plan, doc, or outline of how to order these optimizations and ensure they stay correctly ordered moving forward?

Ideally, we would be able to apply these transformations without needing to consider that `x / pow2` must come before `x * 2.0` because `reason`.

Reply:
no idea, we got lucky these two were executed in a correct order

I guess when you optimize something, e.g. change operator of a node (e.g. `uX GT_GT 0` to `X GT_NE 0`) and feel this particular node can be re-optimized under a different case you should either:
1) `goto case GT_NE_OP` (ugly)
2) call `fgMorphSmpOp` recursively (needs some recursion depth limit just in case)

Reply:
@AndyAyersMS, Do you have any thoughts/comments here?

As we introduce more optimizations, this seems like something that will be easy to miss and it would be good to have a consistent process for how handling it is normally recommended...

Reply:
btw, a good example is this PR: https://github.com/dotnet/coreclr/pull/25744/files

Reply:
What's in morph is pretty messy. We don't have a good way of determining if there are dependencies between optimizations, or if one optimization may enable another. There are long-term ambitions of refactoring all this but it is not something we'll get to anytime soon.

Best I can suggest for now is to make sure you run diffs widely, eg over the entire set of Pri1 tests, and look carefully at the diffs, or (if you are internal) run your change over some of our SPMI collections.

We want to avoid repeated subtree traversals so we don't waste time, get stuck in some oscillating mode or blow the stack; so re-invoking morph recursively is only really viable if the node has been simplified in some way. If you are really sure optimizing A enables optimization B, then factor B out and invoke it as a subroutine as part finishing up A.